core/src/test/groovy/com/novoda/gradle/release/test/TestProjectRule.groovy

```diff
@@ -0,0 +1,75 @@
+package com.novoda.gradle.release.test
+
+
+import org.junit.rules.TestRule
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+
+class TestProjectRule implements TestRule {
+
+    private enum ProjectType {
+        JAVA, ANDROID
+    }
+
+    private final ProjectType project
+    private BuildFolderRule tempFolder
+    private String buildScript
+
+    static TestProjectRule newJavaProject(String buildScript = GradleScriptTemplates.forJavaProject()) {
+        return new TestProjectRule(ProjectType.JAVA, buildScript)
+    }
+
+    static TestProjectRule newAndroidProject(String buildScript = GradleScriptTemplates.forAndroidProject()) {
+        return new TestProjectRule(ProjectType.ANDROID, buildScript)
+    }
+
+    private TestProjectRule(ProjectType project, String buildScript) {
+        this.project = project
+        this.buildScript = buildScript
+    }
+
+    @Override
+    Statement apply(Statement base, Description description) {
+        tempFolder = new BuildFolderRule("test-projects/${description.testClass.canonicalName}/${description.methodName}")
```

Comment:
Does this need to be a rule if we're using it internally in another rule?

Reply:
it was a rule (in another branch) and also this way it helps with the chaining, so we can encapsulate its behaviour in its own statement. The reason why here the rule is created so late is because I wanted to use it in order to create the right folder structure, but that is definitely just something that I want to use in this case, while in another place I could just use the rule on its own and pick something more static